contrib/mypy/src/python/pants/contrib/mypy/tasks/mypy_task.py

```diff
@@ -0,0 +1,136 @@
+    # the target roots or transitive dependencies.)
+    source_roots = set()
+    for target in self.context.targets(self.is_python_target):
+      if not target.has_sources(self._PYTHON_SOURCE_EXTENSION):
+        continue
+      source_roots.add(target.target_base)
+    return source_roots
+
+  def _run_mypy(self, py3_interpreter, mypy_args, **kwargs):
+    pex_info = PexInfo.default()
+    pex_info.entry_point = 'mypy'
+    chroot = self.cached_chroot(interpreter=py3_interpreter, pex_info=pex_info, targets=[],
+      extra_requirements=[PythonRequirement('mypy=={}'.format(self.get_options().mypy_version))])
+    pex = chroot.pex()
+    return pex.run(mypy_args, **kwargs)
+
+  def execute(self):
+    py3_interpreter = self.find_py3_interpreter()
+    if not py3_interpreter:
+      raise TaskError('Unable to find a Python 3.x interpreter (required for mypy).')
+
+    sources = self._calculate_python_sources(self.context.target_roots)
+    if not sources:
+      self.context.log.debug('No Python sources to check.')
+      return
+
+    # Determine interpreter used by the sources so we can tell mypy.
+    interpreter_for_targets = self._interpreter_cache.select_interpreter_for_targets(self.context.target_roots)
+    if not interpreter_for_targets:
+      raise TaskError('No Python interpreter compatible with specified sources.')
+
+    with temporary_file_path() as sources_list_path:
+      with open(sources_list_path, 'w') as f:
+        for source in sources:
+          f.write('{}\n'.format(source))
+
+      # Construct the mypy command line.
+      cmd = ['--python-version={}'.format(interpreter_for_targets.identity.python)]
+      cmd.extend(self.get_passthru_args())
+      cmd.append('@{}'.format(sources_list_path))
+      self.context.log.debug('mypy command: {}'.format(' '.join(cmd)))
+
+      with self.temporary_chroot(interpreter=interpreter_for_targets, pex_info=None,
+        targets=self.context.target_roots, platforms=()) as chroot:
+
+        # Collect source roots for the targets being checked.
+        source_roots = self._collect_source_roots(self.context.targets)
+
+        # Determine the sys.path used for the targets. 
+        script_path = os.path.join(chroot.path(), 'dump-path.py')
+        with open(script_path, 'w') as f:
+          f.write("""import sys ; print(':'.join(sys.path))""")
+        pex = chroot.pex()
+        proc = pex.run([script_path], env={'PEX_INTERPRETER': '1'}, blocking=False,
+          stdout=subprocess.PIPE, stderr=None)
+        stdout, _ = proc.communicate()
+        if proc.returncode != 0:
+          raise TaskError('Unable to determine PYTHONPATH for targets.')
+        pythonpath_for_targets = stdout.strip()
+
+        # Construct the MYPYPATH from both the source roots and targets' sys.path.
+        mypy_path = ':'.join([os.path.join(get_buildroot(), root) for root in source_roots])
+        mypy_path += ':'
+        mypy_path += pythonpath_for_targets
```

Comment:
maybe:

```
mypy_path = ':'.join((mypy_path, pythonpath_for_targets)
```

Reply:
I also could combine the two lines into `mypy_path += ':' + pythonpath_for_targets`. Using `.join` here seems too fancy for plain old string concatenation, unless the style here is to avoid mutability?

Reply:
No strong opinion on join, but it's easy enough to use `os.pathsep` instead of a raw `:`.

Reply:
Good idea. Will do that.